File: data-structures-graph-algorithms-enhanced/src/algorithms/cycleDetection.js
```javascript
const Graph = require('../data_structures/Graph');

/**
 * Detects a cycle in a Directed Graph using DFS.
 * Relies on three states for each node during DFS:
 * 1. Unvisited (not in `visited` or `recursionStack`)
 * 2. Visiting (in `recursionStack`, but not fully processed yet)
 * 3. Visited (in `visited`, fully processed)
 *
 * A cycle is detected if we encounter a node that is currently in the `recursionStack`
 * during a DFS traversal. This indicates a back-edge.
 *
 * @param {Graph} graph - The directed graph instance.
 * @returns {boolean} - True if a cycle is detected, false otherwise.
 *
 * Time Complexity: O(V + E) - Each vertex and each edge is visited once.
 * Space Complexity: O(V) - For `visited` set, `recursionStack` set, and recursion stack of DFS.
 */
function detectCycleDirected(graph) {
    if (!graph.directed) {
        console.warn("detectCycleDirected called on an undirected graph. Please use detectCycleUndirected for undirected graphs.");
        return false; // Or throw an error, depending on desired strictness.
    }

    const visited = new Set();         // Stores all nodes whose DFS traversal has started or completed.
    const recursionStack = new Set();  // Stores nodes currently in the recursion stack (current DFS path).

    // Helper DFS function to check for cycles
    function dfs(vertex) {
        visited.add(vertex);
        recursionStack.add(vertex); // Mark current node as part of the recursion stack

        // Explore neighbors
        for (const neighborInfo of graph.getNeighbors(vertex)) {
            const neighbor = neighborInfo.vertex;

            // If neighbor is in recursionStack, a cycle is detected (back-edge).
            if (recursionStack.has(neighbor)) {
                return true; // Cycle found
            }

            // If neighbor has not been visited, recurse
            if (!visited.has(neighbor)) {
                if (dfs(neighbor)) {
                    return true; // Propagate cycle detection upwards
                }
            }
        }

        // After visiting all neighbors and their subtrees, remove from recursion stack.
        // This node's DFS path is complete.
        recursionStack.delete(vertex);
        return false; // No cycle found in this path
    }

    // Iterate over all vertices to ensure all disconnected components are covered.
    for (const vertex of graph.getVertices()) {
        if (!visited.has(vertex)) {
            if (dfs(vertex)) {
                return true; // Cycle found in one of the DFS traversals
            }
        }
    }

    return false; // No cycles found in any component
}

/**
 * Detects a cycle in an Undirected Graph using DFS.
 * For an undirected graph, a cycle is detected if we encounter an already visited vertex
 * that is not the direct parent of the current vertex in the DFS tree.
 *
 * @param {Graph} graph - The undirected graph instance.
 * @returns {boolean} - True if a cycle is detected, false otherwise.
 *
 * Time Complexity: O(V + E) - Each vertex and each edge is visited once.
 * Space Complexity: O(V) - For `visited` set and recursion stack of DFS.
 */
function detectCycleUndirected(graph) {
    if (graph.directed) {
        console.warn("detectCycleUndirected called on a directed graph. Please use detectCycleDirected for directed graphs.");
        return false; // Or throw an error.
    }

    const visited = new Set(); // Stores all nodes whose DFS traversal has completed.

    // Helper DFS function to check for cycles
    // `parent` is crucial here to distinguish back-edges from tree-edges.
    function dfs(vertex, parent) {
        visited.add(vertex); // Mark current node as visited

        // Explore neighbors
        for (const neighborInfo of graph.getNeighbors(vertex)) {
            const neighbor = neighborInfo.vertex;

            // If neighbor is visited AND is not the parent, a cycle is detected.
            // This is a back-edge to an already visited node that's not its immediate parent.
            if (visited.has(neighbor) && neighbor !== parent) {
                return true; // Cycle found
            }

            // If neighbor has not been visited, recurse
            if (!visited.has(neighbor)) {
                if (dfs(neighbor, vertex)) { // Pass current vertex as parent for the recursive call
                    return true; // Propagate cycle detection upwards
                }
            }
        }
        return false; // No cycle found in this path
    }

    // Iterate over all vertices to ensure all disconnected components are covered.
    for (const vertex of graph.getVertices()) {
        if (!visited.has(vertex)) {
            // Start DFS from this unvisited vertex. Parent is null for the root of a DFS tree.
            if (dfs(vertex, null)) {
                return true; // Cycle found in one of the DFS traversals
            }
        }
    }

    return false; // No cycles found in any component
}

module.exports = {
    detectCycleDirected,
    detectCycleUndirected
};
```

File: leetcode-graph-algorithms-solutions-pro/src/main/java/com/example/graph/algorithms/CycleDetection.java
```java
package com.example.graph.algorithms;

import com.example.graph.datastructures.DisjointSet;
import com.example.graph.datastructures.Graph;
import com.example.graph.datastructures.Edge;

import java.util.*;

/**
 * Provides algorithms for detecting cycles in undirected graphs.
 * Two primary approaches are demonstrated: Depth-First Search (DFS) and Disjoint Set Union (DSU).
 */
public class CycleDetection<V> {

    /**
     * Detects a cycle in an undirected graph using Depth-First Search (DFS).
     *
     * <p>Algorithm:</p>
     * <p>1. Keep track of visited nodes and the parent of the current node in the DFS tree.</p>
     * <p>2. For each unvisited neighbor of the current node, recursively call DFS.</p>
     * <p>3. If a visited neighbor is encountered and it is NOT the parent of the current node,
     *    then a cycle is detected.</p>
     * <p>4. This approach works for undirected graphs because an edge (u, v) means v is a neighbor of u
     *    and u is a neighbor of v. We need to ignore the edge that leads back to the parent in the DFS tree
     *    to avoid falsely detecting a cycle.</p>
     *
     * @param graph The undirected graph to check for cycles.
     * @return true if a cycle is detected, false otherwise.
     * @throws IllegalArgumentException if the graph is directed.
     *
     * Time Complexity: O(V + E) where V is the number of vertices and E is the number of edges.
     *                  Each vertex and edge is visited once.
     * Space Complexity: O(V) for the visited set and recursion stack.
     */
    public boolean detectCycleDFS(Graph<V> graph) {
        if (graph.isDirected()) {
            throw new IllegalArgumentException("DFS cycle detection with parent tracking is suitable for UNDIRECTED graphs.");
        }

        Set<V> visited = new HashSet<>();
        // Iterate over all vertices to handle disconnected components
        for (V vertex : graph.getVertices()) {
            if (!visited.contains(vertex)) {
                // Call DFS starting from this unvisited vertex. Pass null as initial parent.
                if (dfs(graph, vertex, null, visited)) {
                    return true; // Cycle detected in this component
                }
            }
        }
        return false; // No cycle found in any component
    }

    /**
     * Recursive helper for DFS-based cycle detection.
     *
     * @param graph   The graph.
     * @param current The current vertex being visited.
     * @param parent  The parent of the current vertex in the DFS traversal path.
     * @param visited Set of visited vertices.
     * @return true if a cycle is detected, false otherwise.
     */
    private boolean dfs(Graph<V> graph, V current, V parent, Set<V> visited) {
        visited.add(current); // Mark current vertex as visited

        // Explore neighbors
        for (V neighbor : graph.getNeighbors(current)) {
            // If neighbor is not visited, recurse
            if (!visited.contains(neighbor)) {
                if (dfs(graph, neighbor, current, visited)) {
                    return true; // Cycle found in recursive call
                }
            } else if (!neighbor.equals(parent)) {
                // If neighbor is visited AND it's not the parent of current, then a cycle is detected.
                // This condition is crucial for undirected graphs to avoid considering
                // the edge that brought us to 'current' from 'parent' as a cycle.
                return true;
            }
        }
        return false; // No cycle found from this vertex
    }

    /**
     * Detects a cycle in an undirected graph using the Disjoint Set Union (DSU) data structure.
     * This approach processes edges one by one. If an edge connects two vertices that are
     * already in the same set (meaning they are already connected), then adding this edge
     * would form a cycle.
     *
     * <p>Algorithm:</p>
     * <p>1. Initialize a DSU structure where each vertex is in its own set.</p>
     * <p>2. Iterate through all edges (u, v) in the graph.</p>
     * <p>3. For each edge, find the representatives (roots) of the sets containing u and v.</p>
     * <p>4. If the representatives are the same, it means u and v are already connected,
     *    and adding this edge creates a cycle. Return true.</p>
     * <p>5. If the representatives are different, union the sets containing u and v.</p>
     * <p>6. If all edges are processed without finding a cycle, return false.</p>
     *
     * @param graph The undirected graph to check for cycles.
     * @return true if a cycle is detected, false otherwise.
     * @throws IllegalArgumentException if the graph is directed.
     *
     * Time Complexity: O(E * α(V)), where E is the number of edges, V is the number of vertices,
     *                  and α is the inverse Ackermann function, which is practically constant.
     *                  Effectively, it's O(E) for practical purposes.
     * Space Complexity: O(V) for the DisjointSet structure.
     */
    public boolean detectCycleDSU(Graph<V> graph) {
        if (graph.isDirected()) {
            throw new IllegalArgumentException("DSU cycle detection is suitable for UNDIRECTED graphs.");
        }

        DisjointSet<V> dsu = new DisjointSet<>(graph.getVertices());

        // Iterate through all edges
        for (Edge<V> edge : graph.getAllEdges()) {
            V u = edge.getSource();
            V v = edge.getDestination();

            // Find representatives of u and v
            V rootU = dsu.find(u);
            V rootV = dsu.find(v);

            // If they are already in the same set, adding this edge creates a cycle
            if (rootU.equals(rootV)) {
                return true;
            }

            // Otherwise, union their sets
            dsu.union(u, v);
        }
        return false; // No cycle found
    }
}
```